students/ginafs/session05/mailroom3.py

```diff
@@ -0,0 +1,222 @@
+    except KeyboardInterrupt:
+        print_goodbye()
+        return False
+
+    #check if the donor is in the database or is a new donor
+    #if it is a new donor set the new_donor flag to true
+    new_donor = False
+    if donor_name not in DONATIONS:
+        new_donor = True
+
+    ''' exit the program with a message if user enters ^C, or
+    keep prompting the user for a valid input until user enters a
+    valid input. Break out of the loop if user enters valid number'''
+    while True:
+        try:
+            donation_amount = float(input(f"Enter donation amount > "))
+        except KeyboardInterrupt:
+            print_goodbye()
+            return False
+        except ValueError:
+            print("Please enter a valid number for the donation amount")
+            continue
+        else:
+            break
+
+    if new_donor:
+        #add donor_name and donation_amount to donations
+        DONATIONS[donor_name] = [donation_amount]
+    else:
+        #add donation_amount to list of donations of the donor in donations
+        DONATIONS[donor_name].append(donation_amount)
+    print(f"\nThe donation of ${donation_amount} by {donor_name} was successfully added to the database\nThe letter follows\n\n")
+    out_tuple = (donor_name, float(donation_amount))
+    print(SINGLE_DONATION_LETTER.format(*out_tuple))
+    return True
+
+
+def create_report():
+    '''Pretty much the same from Part 1'''
+    print_donor_report()
+    print()
+    print()
+    return True
+
+
+def letters_to_donors():
+    '''Uses the donations dictionary to create Thank you letter.txt files for each
+    donor in donations dict. Builds a file_name from the donor's name and timestamp.
+    Creates modified letters for donors with a single donation and for donors
+    with multiple donations '''
+
+    print("Printing letters to donors")
+    print()
+    for key in DONATIONS:
+        #form file name from key
+        file_name = str(key).replace(' ', '_')
+
+        #get time stamp
+        timestamp = str(datetime.datetime.now()).split('.')[0]
+
+        #format the timestamp into an valid and readable string
+        timestamp = str(timestamp).replace(' ', '_')
+        timestamp = str(timestamp).replace(':', '.')
+        timestamp = str(timestamp).replace('-', '.')
+        file_name += '.' + timestamp
+
+        #get the number of donations for the donor and the total donation amount
+        num_donations = len(DONATIONS[key])
+        total_amount = sum(DONATIONS[key])
+
+        #Build the string to write into the file. Use tuples to send in values
+        out_str = ""
+        if num_donations > 1:
+            out_str = MULTIPLE_DONATION_LETTER
+            out_tuple = (key, int(num_donations), float(total_amount))
+        else:
+            #donor with single donation does not need number of donations
+            out_str = SINGLE_DONATION_LETTER
+            out_tuple = (key, float(total_amount))
```

Comment:
variable names call `out_str` and `out_tuple` don't describe what the data is in those, just a type.